examples/contrastive/contrastive.cpp

```diff
@@ -0,0 +1,223 @@
+
+    fprintf(stderr, "\n");
+
+    for (auto id : tokens_list) {
+        fprintf(stderr, "%s", llama_token_to_piece(ctx_expert, id).c_str());
+    }
+
+    fflush(stderr);
+
+    // create a llama_batch with size 512
+    // we use this object to submit token data for decoding
+
+    llama_batch batch = llama_batch_init(512, 0, 1);
+
+    // evaluate the initial prompt
+    for (size_t i = 0; i < tokens_list.size(); i++) {
+        llama_batch_add(batch, tokens_list[i], i, { 0 }, false);
+    }
+
+    // llama_decode will output logits only for the last token of the prompt
+    batch.logits[batch.n_tokens - 1] = true;
+
+    if (llama_decode(ctx_expert, batch) != 0) {
+        LOG_TEE("%s: llama_decode() failed\n", __func__);
+        return 1;
+    }
+
+    if (llama_decode(ctx_amateur, batch) != 0) {
+        LOG_TEE("%s: llama_decode() failed\n", __func__);
+        return 1;
+    }
+
+    // main loop
+
+    int n_cur    = batch.n_tokens;
+    int n_decode = 0;
+
+    const auto t_main_start = ggml_time_us();
+
+    float alpha = 0.1;
+    float beta = 0.5;
+
+    while (n_cur <= n_len) {
+        // sample the next token
+        {
+            auto   n_vocab = llama_n_vocab(model_expert);
+            auto * logits_expert  = llama_get_logits_ith(ctx_expert, batch.n_tokens - 1);
+            auto * logits_amateur  = llama_get_logits_ith(ctx_amateur, batch.n_tokens - 1);
+
+            std::vector<llama_token_data> candidates;
+            candidates.reserve(n_vocab);
+
+            for (llama_token token_id = 0; token_id < n_vocab; token_id++) {
+                float cd_logit = std::numeric_limits<float>::lowest();
+                if(logits_expert[token_id] > alpha){
+                    cd_logit = (1+beta)*logits_expert[token_id] - beta*logits_amateur[token_id];
+                }
+                candidates.emplace_back(llama_token_data{ token_id, cd_logit, 0.0f });
+            }
+
+            llama_token_data_array candidates_p = { candidates.data(), candidates.size(), false };
+
+            // sample the most likely token
+            const llama_token new_token_id_expert = llama_sample_token_greedy(ctx_expert, &candidates_p);
+
+            // is it an end of stream?
+            if (new_token_id_expert == llama_token_eos(model_expert) || n_cur == n_len) {
+                LOG_TEE("\n");
+
+                break;
+            }
+
+            LOG_TEE("%s", llama_token_to_piece(ctx_expert, new_token_id_expert).c_str());
+            fflush(stdout);
+
+            // prepare the next batch
+            llama_batch_clear(batch);
+
+            // push this new token for next evaluation
+            llama_batch_add(batch, new_token_id_expert, n_cur, { 0 }, true);
+
+            n_decode += 1;
+        }
+
+        n_cur += 1;
+
+        // evaluate the current batch with the transformer model
+        if (llama_decode(ctx_expert, batch)) {
+            fprintf(stderr, "%s : failed to eval, return code %d\n", __func__, 1);
```

Comment:
Why have a format argument if it's a constant?

Reply:
erased format argument.